### DISCUSSION TITLE
General computation of diffusion matrices

### DISCUSSION THREAD

**Discussion by swhite2401** [atcollab/at#715] [Ideas]
from(@lfarv) @swhite2401, @simoneliuzzo : here is an idea of how we could progress in computing the equilibrium emittance in a more general way. I'm not yet sure it works, but I would like your comments on the idea:

Add a new entry point in the concerned integrators: this new entry point has the same input arguments as trackFunction: particle coordinates and element description. It has an additional output argument, returning the diffusion matrix. Internally, this will call a single integrator function with a flag triggering the computation of diffusion if necessary. This can be tested in BndMPoleSymplectic4RadPass, and then generalised to other integrators. At 1st glance, one could reproduce the diffusion computation for all integrators using a drift-kick sequence: BndMPoleSymplectic4RadPass, StrMPoleSymplectic4RadPass, GWigSymplecticPass. And probably their "Exact" counterparts, to be checked. This development works for both Matlab and python.

In ohmi_envelope, replace findmpoleraddifmatrix by a kind of copy of at.c (python) or atpass.c (Matlab). It deals with a single element (not the whole line as at.c), and has to load the required shared libraries, build the table of entry points, and call the one selected by the element. And returns the diffusion matrix !

Merging what's in findmpoleraddifmatrix and in BndMPoleSymplectic4RadPass is not easy, but looks possible.

**Comment by simoneliuzzo** [comment]
I have not enough knowledge on the trackFunction and .c code. 
However I would really appreciate such upgrade. For example using Exact---Bend---  integrators still makes emittances NaN from ATX.

**Comment by lfarv** [comment]
Thanks for extracting this comment and moving it into discussions. It had been lost in PR comments…

I think the proposal is still valid, but I did not spend any time on that.

However, I'm surprised by @simoneliuzzo's problem: in my tests, ExactSectorBendRadPass gives correct damping times and emittances. Though it does not use the exact passmethod to compute the diffusion matrix, as previously mentioned.

**Comment by carmignani** [comment]
Hello,
I think this is one of the most annoying old AT problem and I am in favour of any solution to get rid of findmpoleraddifmatrix.
I had a quick look to the code to remember how it works. 
The new entry function for the passmethods could be a solution, but maybe it is not even needed: the passMethod could have an additional input `double *BDIFF` that if is not empty triggers the computation of the diffusion matrix. We would necessarily have a lot of checks if `BDIFF` is empty during the tracking, but I don't think it is avoidable.

**Comment by lfarv** [comment]
Hi all
I like @carmignani's proposal avoiding a new entry point in the integrators. The problem is that user-written integrators which would not be updated could crash in an obscure way: the way incompatible argument lists in C are treated depends on the compiler. So there is the possibility of putting the "double *" argument in a new slot of the "struct parameters" existing argument. Not very elegant, but it avoids all compatibility problems.

**Comment by carmignani** [comment]
I agree about putting the `double *` in the struct parameters. If there are no volunteers to do this development sooner, I could try to make a proposal for BndMPoleSymplectic4RadPass and then in case we extend to the Str version and the others. I'm not very efficient to write python code, but I like C! I think I cannot do this before mid March.

**Comment by lfarv** [comment]
Hi @carmignani: sorry to intervene in your plan… I started to look at that as soon as @swhite2401 revived the topic, and adopted your proposal of keeping a single entry point in integrators. So at the moment, I have BndMPoleSymplectic4Pass and StrMPoleSymplectic4Pass working, with the python interface. The Matlab interface is a bit more delicate, I'm now working on it. It will take some more time for the "exact" passmethods since we have to derive the correct formalae, but I see how it will work.

**Comment by ZeusMarti** [comment]
Hi @lfarv and @carmignani, we realized GWigSymplecticRadPass does not contribute in the present ohmienvelope version but I guess it would be nice to add it in this new strategy. Are you planning to work on it as well? Maybe @mashl? Otherwise we are starting to review the topic student, so let us know if you want some help!
